src/main/go/AlexanderYastrebov/calc.go

```diff
@@ -0,0 +1,282 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"runtime"
+	"sort"
+	"sync"
+	"syscall"
+)
+
+type measurement struct {
+	min, max, sum, count int64
+}
+
+func main() {
+	if len(os.Args) != 2 {
+		log.Fatalf("Missing measurements filename")
+	}
+
+	measurements := processFile(os.Args[1])
+
+	ids := make([]string, 0, len(measurements))
+	for id := range measurements {
+		ids = append(ids, id)
+	}
+	sort.Strings(ids)
+
+	fmt.Print("{")
+	for i, id := range ids {
+		if i > 0 {
+			fmt.Print(", ")
+		}
+		m := measurements[id]
+		fmt.Printf("%s=%.1f/%.1f/%.1f", id, round(float64(m.min)/10.0), round(float64(m.sum)/10.0/float64(m.count)), round(float64(m.max)/10.0))
+	}
+	fmt.Println("}")
+}
+
+func processFile(filename string) map[string]*measurement {
+	f, err := os.Open(filename)
+	if err != nil {
+		log.Fatalf("Open: %v", err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatalf("Stat: %v", err)
+	}
+
+	size := fi.Size()
+	if size <= 0 || size != int64(int(size)) {
+		log.Fatalf("Invalid file size: %d", size)
+	}
+
+	data, err := syscall.Mmap(int(f.Fd()), 0, int(size), syscall.PROT_READ, syscall.MAP_SHARED)
+	if err != nil {
+		log.Fatalf("Mmap: %v", err)
+	}
+
+	defer func() {
+		if err := syscall.Munmap(data); err != nil {
+			log.Fatalf("Munmap: %v", err)
+		}
+	}()
+
+	return process(data)
+}
+
+func process(data []byte) map[string]*measurement {
+	nChunks := runtime.NumCPU()
+
+	chunkSize := len(data) / nChunks
+	if chunkSize == 0 {
+		chunkSize = len(data)
+	}
+
+	chunks := make([]int, 0, nChunks)
+	offset := 0
+	for offset < len(data) {
+		offset += chunkSize
+		if offset >= len(data) {
+			chunks = append(chunks, len(data))
+			break
+		}
+
+		nlPos := bytes.IndexByte(data[offset:], '\n')
+		if nlPos == -1 {
+			chunks = append(chunks, len(data))
+			break
+		} else {
+			offset += nlPos + 1
+			chunks = append(chunks, offset)
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(chunks))
+
+	results := make([]map[string]*measurement, len(chunks))
+	start := 0
+	for i, chunk := range chunks {
+		go func(data []byte, i int) {
+			results[i] = processChunk(data)
+			wg.Done()
+		}(data[start:chunk], i)
+		start = chunk
+	}
+	wg.Wait()
+
+	measurements := make(map[string]*measurement)
+	for _, r := range results {
+		for id, rm := range r {
+			m := measurements[id]
+			if m == nil {
+				measurements[id] = rm
+			} else {
+				m.min = min(m.min, rm.min)
+				m.max = max(m.max, rm.max)
+				m.sum += rm.sum
+				m.count += rm.count
+			}
+		}
+	}
+	return measurements
+}
+
+func processChunk(data []byte) map[string]*measurement {
+	// Use fixed size linear probe lookup table
+	const (
+		// use power of 2 for fast modulo calculation
+		entriesSize = 1 << 12
```

Comment:
Hm, this should not pass 10k keys test, need to check.

Reply:
> Hm, this should not pass 10k keys test but it does, need to check.

Maybe it passed because each of your chunks did not have more than 4096 cities?